src/lib/actions-watcher.js

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2024 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+const upath = require('upath')
+const chokidar = require('chokidar')
+const coreLogger = require('@adobe/aio-lib-core-logging')
+const { buildActions } = require('@adobe/aio-lib-runtime')
+const watchLogger = coreLogger('watcher', { level: process.env.LOG_LEVEL, provider: 'winston' })
+
+/**
+ * @typedef {object} WatchReturnObject
+ * @property {object} watcher the watcher object
+ * @property {Function} cleanup callback function to cleanup available resources
+ */
+
+/**
+ * @typedef {object} WatcherOptions
+ * @property {object} config the app config (see src/lib/config-loader.js)
+ * @property {boolean} isLocal whether the deployment is local or not
+ * @property {Function} log the app logger
+ * @property {object} [watcher] the watcher itself
+ */
+
+/**
+ * Create a watcher.
+ *
+ * @param {WatcherOptions} watcherOptions the options for the watcher
+ * @returns {WatchReturnObject} the WatchReturnObject
+ */
+async function createWatcher (watcherOptions) {
+  const { config } = watcherOptions
+
+  watchLogger.info(`watching action files at ${config.actions.src}...`)
+  const watcher = chokidar.watch(config.actions.src)
+  const onChangeHandler = createChangeHandler({ ...watcherOptions, watcher })
+
+  watcher.on('change', onChangeHandler)
+
+  const watcherCleanup = async () => {
+    watchLogger.debug('stopping action watcher...')
+    await watcher.close()
+  }
+
+  return {
+    onChangeHandler,
+    watcher,
+    watcherCleanup
+  }
+}
+
+/**
+ * Simple queue
+ */
+class Queue {
```

Comment:
I would argue that after seeing the previous action-watcher code in the app-plugin which was hard to parse how it works, there is a major benefit in making the syntax clear in this regard for maintainability purposes (versus .shift() etc which is quite unclear at a glance without going to the docs, since it mutates an array etc). 

Especially since this is not critical path code where a  performance improvement (negligible perhaps, I didn't do a profile) is needed.